runtime-light/coroutine/shared-task.h

```diff
@@ -0,0 +1,361 @@
+    const void *const NOT_STARTED_VAL{std::addressof(this->m_waiters)};
+
+    // NOTE: If the coroutine is not yet started then the first waiter
+    // will start the coroutine before enqueuing itself up to the list
+    // of suspended waiters waiting for completion. We split this into
+    // two steps to allow the first awaiter to return without suspending.
+    // This avoids recursively resuming the first waiter inside the call to
+    // coroutine.resume() in the case that the coroutine completes
+    // synchronously, which could otherwise lead to stack-overflow if
+    // the awaiting coroutine awaited many synchronously-completing
+    // tasks in a row.
+
+    // start the coroutine if not yet started
+    if (m_waiters == NOT_STARTED_VAL) {
+      m_waiters = STARTED_NO_WAITERS_VAL;
+      std::coroutine_handle<promise_type>::from_promise(*static_cast<promise_type *>(this)).resume();
+    }
+    // coroutine already completed, don't suspend
+    if (ready()) {
+      return false;
+    }
+
+    waiter.m_next = static_cast<shared_task_impl_::shared_task_waiter_t *>(m_waiters);
+    m_waiters = static_cast<void *>(std::addressof(waiter));
+    return true;
+  }
+
+  // return true if successfully detached, false if this was the last
+  // reference to the coroutine, in which case the caller must
+  // call destroy() on the coroutine handle.
+  constexpr bool detach() noexcept {
+    return m_refcnt-- != 1;
+  }
+
+  void cancel_awaiter(const shared_task_impl_::shared_task_waiter_t &waiter) noexcept {
+    const void *const READY_VAL{this};
+    if (m_waiters == READY_VAL) [[unlikely]] {
+      php_critical_error("currently, shared_task_t does not support cancellation after it has finished");
+    }
+
+    const void *const NOT_STARTED_VAL{std::addressof(this->m_waiters)};
+    if (m_waiters == NOT_STARTED_VAL || m_waiters == STARTED_NO_WAITERS_VAL) [[unlikely]] {
+      return;
+    }
+
+    const auto *waiter_ptr{std::addressof(waiter)};
+    if (m_waiters == waiter_ptr) { // waiter is the head of the list
+      m_waiters = waiter_ptr->m_next;
+      return;
+    }
+
+    auto *tmp_waiters{static_cast<shared_task_impl_::shared_task_waiter_t *>(m_waiters)};
+    while (tmp_waiters->m_next != nullptr && tmp_waiters->m_next != waiter_ptr) {
+      tmp_waiters = tmp_waiters->m_next;
+    }
+    if (tmp_waiters->m_next == nullptr) [[unlikely]] { // for some reason there is no such waiter
+      return;
+    }
+
+    tmp_waiters->m_next = tmp_waiters->m_next->m_next;
+  }
+
+  template<typename... Args>
+  void *operator new(std::size_t n, [[maybe_unused]] Args &&...args) noexcept {
+    // todo:k2 think about args in new
+    // todo:k2 make coroutine allocator
+    void *buffer = k2::alloc(n);
+    return buffer;
+  }
+
+  void operator delete(void *ptr, [[maybe_unused]] size_t n) noexcept {
+    k2::free(ptr);
+  }
+
+private:
+  static constexpr void *STARTED_NO_WAITERS_VAL = nullptr;
+
+  uint32_t m_refcnt{1};
```

Comment:
If we want to optimize size of promise_base_t, what do you think about reordering struct fields and enabling something like `#pragma pack`

Reply:
And maybe add some static asserts on structs size?